### ISSUE TITLE
Type redefinition errors when trying to include "sys/types.h"

### ISSUE THREAD

**Issue by adiemu5** [iovisor/bpftrace#1511]
### bpftrace --info
System
  OS: Linux 3.10.0-1127.19.1.el7.x86_64 #1 SMP Tue Aug 25 17:23:54 UTC 2020
  Arch: x86_64

Build
  version: v0.11.0
  LLVM: 8
  foreach_sym: yes
  unsafe uprobe: no
  bfd: yes
  bpf_attach_kfunc: yes
  bcc_usdt_addsem: no
  libbpf: no
  libbpf btf dump: no
  libbpf btf dump type decl: no

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: yes

Kernel features
  Instruction limit: -1
  Loop support: no
  btf: no

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes

Probe types
  kprobe: no
  tracepoint: yes
  perf_event: yes
  kfunc: no

NOTE: Using the bpftrace RPMs from https://github.com/fbs/el7-bpf-specs on CentOS 7

### What reproduces the bug?
Trying to do some purely usermode tracing, and need to #include <sys/types.h> but get a ton of redefinition errors because it seems bpftrace includes <linux/types.h> by default, as mentioned in #876  .

Is there a way to turn off bpftrace's default includes?  They seem to presume kernel tracing in a way that prevents deeper usermode tracing.

The simplest reproducer I've found is actually part of bpftrace's test suite, added in 32dd14b:

```
# bpftrace -e "$(echo "#include <sys/types.h>"; echo "BEGIN { exit(); }")"
/usr/include/sys/types.h:44:18: error: typedef redefinition with different types ('__loff_t' (aka 'long') vs '__kernel_loff_t' (aka 'long long'))
/lib/modules/3.10.0-1127.19.1.el7.x86_64/source/include/linux/types.h:45:26: note: previous definition is here
/usr/include/sys/types.h:60:17: error: typedef redefinition with different types ('__dev_t' (aka 'unsigned long') vs '__kernel_dev_t' (aka 'unsigned int'))
/lib/modules/3.10.0-1127.19.1.el7.x86_64/source/include/linux/types.h:15:25: note: previous definition is here
/usr/include/sys/types.h:75:19: error: typedef redefinition with different types ('__nlink_t' (aka 'unsigned long') vs '__u32' (aka 'unsigned int'))
/lib/modules/3.10.0-1127.19.1.el7.x86_64/source/include/linux/types.h:19:17: note: previous definition is here
/usr/include/time.h:103:19: error: typedef redefinition with different types ('__timer_t' (aka 'void *') vs '__kernel_timer_t' (aka 'int'))
/lib/modules/3.10.0-1127.19.1.el7.x86_64/source/include/linux/types.h:25:26: note: previous definition is here
/usr/include/sys/types.h:197:1: error: typedef redefinition with different types ('long' vs '__s64' (aka 'long long'))
/lib/modules/3.10.0-1127.19.1.el7.x86_64/source/include/linux/types.h:113:17: note: previous definition is here
/usr/include/sys/types.h:203:1: error: typedef redefinition with different types ('unsigned long' vs '__u64' (aka 'unsigned long long'))
/lib/modules/3.10.0-1127.19.1.el7.x86_64/source/include/linux/types.h:112:17: note: previous definition is here
/usr/include/sys/select.h:75:5: error: typedef redefinition with different types ('struct fd_set' vs '__kernel_fd_set')
/lib/modules/3.10.0-1127.19.1.el7.x86_64/source/include/linux/types.h:14:26: note: previous definition is here
/usr/include/sys/types.h:235:20: error: typedef redefinition with different types ('__blkcnt_t' (aka 'long') vs 'unsigned long')
/lib/modules/3.10.0-1127.19.1.el7.x86_64/source/include/linux/types.h:134:23: note: previous definition is here
```

**Comment by danobi** [comment]
Hmm, can't seem to reproduce on master:
```
$ sudo ./build/src/bpftrace ~/scratch/bpftrace/bugs/sys_types.h
Attaching 1 probe...

$ sudo ./build/src/bpftrace ~/scratch/bpftrace/bugs/sys_types.h --btf
Attaching 1 probe...

$ cat ~/scratch/bpftrace/bugs/sys_types.h
#include <sys/types.h>

BEGIN
{
  exit();
}
```

**Comment by fbs** [comment]
So, after some experiments, I discovered a problem with this approach - when asking BTF to dump a type definition, it also dumps definitions of all other types needed for successful compilation. If one of those types is overridden by the user, a re-definition error occurs (and it's quite difficult to discover which types were dumped).

This leads me to another thought - do we really want to allow user to do this? I'd say that allowing overriding a part of kernel types while taking the rest from BTF will always eventually blow up somewhere. I'd suggest enabling only the following scenarios:

    there are no user-defined types -> use BTF
    there are user-defined types but they do not override kernel types -> use BTF
    there are user-defined types conflicting with kernel types -> don't use BTF

This should allow user to define his own types that use existing kernel types without any problems, but if he wants to use types with conflicting names (e.g. for userspace tracing), he must define/include all the necessary types manually (or use them from linux/types.h, which is included in such case). This approach should be also backwards compatible for scripts that include kernel headers (such as the above tcplife.bt) as those must have already specified all the necessary types.

Any thoughs before I start implementing this?

Originally posted by @viktormalik in #1506 (comment)

**Comment by fbs** [comment]
#1588 probably will have a lot  of overlap with this

**Comment by woodgear** [comment]
reproduce this error in the latest master build. is this been fixed or not?

**Comment by viktormalik** [comment]
I cannot reproduce this anymore, closing.

Please reopen if you find a reproducer with the current master.

**Comment by honghaier250** [comment]
```shell
# cat ngx_http_process_request_uri.bt

#!/usr/bin/env bpftrace

#include <ngx_core.h>
#include <ngx_http.h>
#include <ngx_config.h>

uprobe:/usr/local/nginx/sbin/nginx:ngx_http_process_request_uri {
    $r = arg0;
    printf("get url:%s\n", str($r->uri));
}
```

[nginx.zip](https://github.com/user-attachments/files/17614589/nginx.zip)


```shell
# sudo bpftrace -I nginx/ ngx_http_process_request_uri.bt

/lib/modules/6.11.2-amd64/source/include/linux/compiler-clang.h:45:9: warning: '__HAVE_BUILTIN_BSWAP32__' macro redefined [-Wmacro-redefined]
note: previous definition is here
/lib/modules/6.11.2-amd64/source/include/linux/compiler-clang.h:46:9: warning: '__HAVE_BUILTIN_BSWAP64__' macro redefined [-Wmacro-redefined]
note: previous definition is here
/lib/modules/6.11.2-amd64/source/include/linux/compiler-clang.h:47:9: warning: '__HAVE_BUILTIN_BSWAP16__' macro redefined [-Wmacro-redefined]
note: previous definition is here
/usr/include/x86_64-linux-gnu/sys/types.h:42:18: error: typedef redefinition with different types ('__loff_t' (aka 'long') vs '__kernel_loff_t' (aka 'long long'))
/lib/modules/6.11.2-amd64/source/include/linux/types.h:52:26: note: previous definition is here
/usr/include/x86_64-linux-gnu/sys/types.h:59:17: error: typedef redefinition with different types ('__dev_t' (aka 'unsigned long') vs '__kernel_dev_t' (aka 'unsigned int'))
/lib/modules/6.11.2-amd64/source/include/linux/types.h:21:25: note: previous definition is here
/usr/include/x86_64-linux-gnu/sys/types.h:74:19: error: typedef redefinition with different types ('__nlink_t' (aka 'unsigned long') vs 'u32' (aka 'unsigned int'))
/lib/modules/6.11.2-amd64/source/include/linux/types.h:25:15: note: previous definition is here
/usr/include/x86_64-linux-gnu/bits/types/timer_t.h:7:19: error: typedef redefinition with different types ('__timer_t' (aka 'void *') vs '__kernel_timer_t' (aka 'int'))
/lib/modules/6.11.2-amd64/source/include/linux/types.h:31:26: note: previous definition is here
/usr/include/x86_64-linux-gnu/bits/stdint-intn.h:27:19: error: typedef redefinition with different types ('__int64_t' (aka 'long') vs 's64' (aka 'long long'))
/lib/modules/6.11.2-amd64/source/include/linux/types.h:115:15: note: previous definition is here
/usr/include/x86_64-linux-gnu/sys/types.h:161:20: error: typedef redefinition with different types ('__uint64_t' (aka 'unsigned long') vs 'u64' (aka 'unsigned long long'))
/lib/modules/6.11.2-amd64/source/include/linux/types.h:114:15: note: previous definition is here
/usr/include/x86_64-linux-gnu/sys/select.h:70:5: error: typedef redefinition with different types ('struct fd_set' vs '__kernel_fd_set')
/lib/modules/6.11.2-amd64/source/include/linux/types.h:20:26: note: previous definition is here
/usr/include/x86_64-linux-gnu/sys/types.h:205:22: error: typedef redefinition with different types ('__blkcnt64_t' (aka 'long') vs 'u64' (aka 'unsigned long long'))
/lib/modules/6.11.2-amd64/source/include/linux/types.h:135:13: note: previous definition is here
/lib/modules/6.11.2-amd64/source/arch/x86/include/uapi/asm/stat.h:83:8: error: redefinition of 'stat'
/usr/include/x86_64-linux-gnu/bits/struct_stat.h:26:8: note: previous definition is here
/lib/modules/6.11.2-amd64/source/arch/x86/include/uapi/asm/stat.h:97:19: error: expected ';' at end of declaration list
/lib/modules/6.11.2-amd64/source/arch/x86/include/uapi/asm/stat.h:99:19: error: expected ';' at end of declaration list
/lib/modules/6.11.2-amd64/source/arch/x86/include/uapi/asm/stat.h:101:19: error: expected ';' at end of declaration list
/lib/modules/6.11.2-amd64/source/arch/x86/include/uapi/asm/stat.h:132:16: error: expected ';' at end of declaration list
/lib/modules/6.11.2-amd64/source/arch/x86/include/uapi/asm/stat.h:133:16: error: expected ';' at end of declaration list
/lib/modules/6.11.2-amd64/source/arch/x86/include/uapi/asm/stat.h:134:16: error: expected ';' at end of declaration list
/lib/modules/6.11.2-amd64/source/include/uapi/linux/time.h:33:8: error: redefinition of 'timezone'
/usr/include/x86_64-linux-gnu/sys/time.h:52:8: note: previous definition is here
/lib/modules/6.11.2-amd64/source/arch/x86/include/uapi/asm/sigcontext.h:40:8: error: redefinition of '_fpx_sw_bytes'
/usr/include/x86_64-linux-gnu/bits/sigcontext.h:31:8: note: previous definition is here
/lib/modules/6.11.2-amd64/source/arch/x86/include/uapi/asm/sigcontext.h:85:8: error: redefinition of '_fpreg'
/usr/include/x86_64-linux-gnu/bits/sigcontext.h:40:8: note: previous definition is here
/lib/modules/6.11.2-amd64/source/arch/x86/include/uapi/asm/sigcontext.h:91:8: error: redefinition of '_fpxreg'
/usr/include/x86_64-linux-gnu/bits/sigcontext.h:46:8: note: previous definition is here
fatal error: too many errors emitted, stopping now [-ferror-limit=]
WARNING: Cannot take type definitions from BTF since there is a redefinition conflict with user-defined types.
```

**Comment by alexcriss** [comment]
I am still seeing this on the lates released version

```
./bpftrace --version
bpftrace v0.22.1
```

Trying to trace `nginx` but getting conflicts

```
./bpftrace -I ./nginx/objs/ -I nginx/src/core/ -I ./nginx/src/http/ -I ./nginx/src/event/ -I ./nginx/src/os/unix/ -I /usr/include/x86_64-linux-gnu ale.bt
/lib/modules/5.10.0-30-amd64/source/include/linux/compiler-clang.h:51:9: warning: '__HAVE_BUILTIN_BSWAP32__' macro redefined [-Wmacro-redefined]
note: previous definition is here
/lib/modules/5.10.0-30-amd64/source/include/linux/compiler-clang.h:52:9: warning: '__HAVE_BUILTIN_BSWAP64__' macro redefined [-Wmacro-redefined]
note: previous definition is here
/lib/modules/5.10.0-30-amd64/source/include/linux/compiler-clang.h:53:9: warning: '__HAVE_BUILTIN_BSWAP16__' macro redefined [-Wmacro-redefined]
note: previous definition is here
/usr/include/x86_64-linux-gnu/sys/types.h:42:18: error: typedef redefinition with different types ('__loff_t' (aka 'long') vs '__kernel_loff_t' (aka 'long long'))
/lib/modules/5.10.0-30-amd64/source/include/linux/types.h:46:26: note: previous definition is here
/usr/include/x86_64-linux-gnu/sys/types.h:59:17: error: typedef redefinition with different types ('__dev_t' (aka 'unsigned long') vs '__kernel_dev_t' (aka 'unsigned int'))
/lib/modules/5.10.0-30-amd64/source/include/linux/types.h:16:25: note: previous definition is here
/usr/include/x86_64-linux-gnu/sys/types.h:74:19: error: typedef redefinition with different types ('__nlink_t' (aka 'unsigned long') vs 'u32' (aka 'unsigned int'))
/lib/modules/5.10.0-30-amd64/source/include/linux/types.h:20:15: note: previous definition is here
/usr/include/x86_64-linux-gnu/bits/types/timer_t.h:7:19: error: typedef redefinition with different types ('__timer_t' (aka 'void *') vs '__kernel_timer_t' (aka 'int'))
/lib/modules/5.10.0-30-amd64/source/include/linux/types.h:26:26: note: previous definition is here
/usr/include/x86_64-linux-gnu/bits/stdint-intn.h:27:19: error: typedef redefinition with different types ('__int64_t' (aka 'long') vs 's64' (aka 'long long'))
/lib/modules/5.10.0-30-amd64/source/include/linux/types.h:109:15: note: previous definition is here
/usr/include/x86_64-linux-gnu/sys/types.h:161:20: error: typedef redefinition with different types ('__uint64_t' (aka 'unsigned long') vs 'u64' (aka 'unsigned long long'))
/lib/modules/5.10.0-30-amd64/source/include/linux/types.h:108:15: note: previous definition is here
/usr/include/x86_64-linux-gnu/sys/select.h:70:5: error: typedef redefinition with different types ('struct fd_set' vs '__kernel_fd_set')
/lib/modules/5.10.0-30-amd64/source/include/linux/types.h:15:26: note: previous definition is here
/usr/include/x86_64-linux-gnu/sys/types.h:205:22: error: typedef redefinition with different types ('__blkcnt64_t' (aka 'long') vs 'u64' (aka 'unsigned long long'))
/lib/modules/5.10.0-30-amd64/source/include/linux/types.h:126:13: note: previous definition is here
/lib/modules/5.10.0-30-amd64/source/arch/x86/include/uapi/asm/stat.h:83:8: error: redefinition of 'stat'
/usr/include/x86_64-linux-gnu/bits/stat.h:46:8: note: previous definition is here
/usr/include/x86_64-linux-gnu/bits/stat.h:94:26: error: expected ';' at end of declaration list
/usr/include/x86_64-linux-gnu/bits/stat.h:95:26: error: expected ';' at end of declaration list
/usr/include/x86_64-linux-gnu/bits/stat.h:96:26: error: expected ';' at end of declaration list
/usr/include/x86_64-linux-gnu/bits/stat.h:94:26: error: expected ';' at end of declaration list
/usr/include/x86_64-linux-gnu/bits/stat.h:95:26: error: expected ';' at end of declaration list
/usr/include/x86_64-linux-gnu/bits/stat.h:96:26: error: expected ';' at end of declaration list
/lib/modules/5.10.0-30-amd64/source/include/uapi/linux/time.h:33:8: error: redefinition of 'timezone'
/usr/include/x86_64-linux-gnu/sys/time.h:52:8: note: previous definition is here
/lib/modules/5.10.0-30-amd64/source/arch/x86/include/asm/bitops.h:283:28: error: static declaration of 'ffs' follows non-static declaration
/usr/include/strings.h:104:12: note: previous declaration is here
/lib/modules/5.10.0-30-amd64/source/arch/x86/include/uapi/asm/sigcontext.h:40:8: error: redefinition of '_fpx_sw_bytes'
/usr/include/x86_64-linux-gnu/bits/sigcontext.h:31:8: note: previous definition is here
/lib/modules/5.10.0-30-amd64/source/arch/x86/include/uapi/asm/sigcontext.h:85:8: error: redefinition of '_fpreg'
/usr/include/x86_64-linux-gnu/bits/sigcontext.h:40:8: note: previous definition is here
fatal error: too many errors emitted, stopping now [-ferror-limit=]
WARNING: Cannot take type definitions from BTF since there is a redefinition conflict with user-defined types.
```

Is there way to work around these errors?

**Comment by viktormalik** [comment]
@alexcriss this looks like a conflict between `linux/types.h` and `sys/types.h` which should only happen if the system doesn't have BTF. Is that your case?

**Comment by alexcriss** [comment]
That seems strange, I have

```
grep CONFIG_DEBUG_INFO_BTF /boot/config-5.10.0-30-amd64
CONFIG_DEBUG_INFO_BTF=y
```

**Comment by viktormalik** [comment]
Hm, then it could be another error that I just found. Would you mind trying #3735 to see if it fixes the problem?

**Comment by viktormalik** [comment]
Ok, can you give us your `bpftrace --info`?

**Comment by viktormalik** [comment]
Both have `btf = yes` so it's not clear to me why `<linux/types.h>` would be included at all (unless some Nginx header includes it). Are you using any tracepoint probes?

**Comment by viktormalik** [comment]
I see. Then I'll try to reproduce, do you have any simple steps how to do it (especially how to get the header files that the script includes)?

**Comment by alexcriss** [comment]
I think this should work
```
mkdir -p /tmp/bpftrace-test
cd /tmp/bpftrace-test

# clone nginx and build auto header files
git clone https://github.com/nginx/nginx.git
cd nginx
# I am not sure if I needed some more dev libraries here, I don't think so, but maybe some packages I have installed are needed for this, besides build-essential
./auto/configure --without-http_gzip_module

# test file
cd ..
cat > test.bt << 'EOF'
#!/usr/bin/bpftrace

#include "./nginx/objs/ngx_auto_headers.h"
#include "./nginx/src/core/ngx_config.h"
#include "./nginx/src/core/ngx_core.h"
#include "./nginx/src/core/ngx_http.h"

uprobe:/usr/local/sbin/nginx:ngx_http_upstream_process_request {
    printf("%s", ((ngx_http_request_t *) arg1)->uri->data);
}
EOF

bpftrace -I ./nginx/objs/ -I nginx/src/core/ -I ./nginx/src/http/ -I ./nginx/src/event/ -I ./nginx/src/os/unix/ -I /usr/include/x86_64-linux-gnu test.bt
```

**Comment by viktormalik** [comment]
Thanks, I was able to reproduce.

The problem seems to be that bpftrace by default adds kernel directories to include paths (`-I/lib/modules/5.10.0-30-amd64/source/include` in your case). Then, some includes in Nginx code apparently make Clang pull headers from both the kernel and the userspace include dirs, causing these conflicts.

I opened #3744 to track this issue, let's continue the discussion there.
